Web Inspector: Glitches when trying to edit a style from a style sheet that has an @import statement

the-chenergy opened the pull request:
#### 227d5b5816c0493843f5ec63f3aaa869c631e2b8
<pre>
Web Inspector: Glitches when trying to edit a style from a style sheet that has an @import statement
<a href="https://rdar.apple.com/125185110">rdar://125185110</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=271403">https://bugs.webkit.org/show_bug.cgi?id=271403</a>

Reviewed by BJ Burg.

This commit optimizes the handling of CSS.setStyleText by using CSSOM
to update the style declaration text directly, to avoid re-parsing the
entire style sheet.

The observed glitch is a combination of two bugs, both due to having to
re-parse for any style edits:

   1. In the backend, when a style sheet with @import statements is
      re-parsed, it (not just the @import-ed ones) may get removed and
      re-added, causing the CSS agent to mistakenly send
      StyleSheetRemoved and StyleSheetAdded events for that parent
      style sheet. (<a href="https://webkit.org/b/279773)">https://webkit.org/b/279773)</a>

   2. Due to @import-ed style sheets receiving new IDs as a result of
      re-parsing, the frontend thinks any edit to the parent style sheet
      is a &quot;significant change.&quot; (Source: <a href="https://github.com/WebKit/WebKit/blob/24b5b087d18d4baa487dd73905c46589fa353da3/Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js#L267)">https://github.com/WebKit/WebKit/blob/24b5b087d18d4baa487dd73905c46589fa353da3/Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js#L267)</a>
      Handling a significant change triggers a re-render of the entire
      styles panel.

Adjacent to this glitch, there&apos;s another notable consequence involving
re-parsing:

   3. Re-parsing an @import-ed style sheet uses its original text from
      its source, meaning editing the parent style sheet will suddenly
      revert all edits made to any @import-ed style sheets.

By avoiding re-parsing for style declaration edits, we fix the immediate
glitch in the simplest fashion and move towards a more efficient
approach to handle editing. This commit is a step in the direction of
eliminating re-parsing for all style editing functions in the inspector:
   1. Editing the style declaration text is addressed by this PR.
   2. Editing the header of a style rule is already done through CSSOM: <a href="https://github.com/WebKit/WebKit/blob/6277bccb19de5ebd68797f369f8ff24fc4417f2e/Source/WebCore/inspector/InspectorStyleSheet.cpp#L1116-L1119.">https://github.com/WebKit/WebKit/blob/6277bccb19de5ebd68797f369f8ff24fc4417f2e/Source/WebCore/inspector/InspectorStyleSheet.cpp#L1116-L1119.</a>
   3. Editing the header of a non-style rule isn&apos;t directly supported by
      CSSOM but can be achieved equivalently by removing and
      re-inserting the updated rule using CSSOM.
   4. Adding and 5. removing rules can be optimized to use CSSOM in a
      similar way to this patch.

(Functions 3, 4, and 5 may still malfunction with @import rules after
this patch due to <a href="https://webkit.org/b/279773">https://webkit.org/b/279773</a> as they&apos;re still done by
re-parsing.)

* Source/WebCore/inspector/InspectorStyleSheet.h:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::computeCanonicalRuleText):
(WebCore::InspectorStyleSheet::setRuleStyleText):
   - Still canonicalize the rule body text to enable showing it in the
     Sources tab, but don&apos;t re-parse the whole style sheet just to apply
     the styles. Use CSSOM&apos;s support for that.

(WebCore::InspectorStyleSheetForInlineStyle::setRuleStyleText):
   - Adapt to the signature change of setRuleStyleText.

(WebCore::InspectorStyleSheet::styleSheetTextWithChangedStyle): Deleted.
   - Unused.

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
   - Add field to record the old rule body text to enable faithful
     undoing.

* LayoutTests/inspector/css/setStyleText-expected.txt: Added.
* LayoutTests/inspector/css/setStyleText.html: Added.

Canonical link: <a href="https://commits.webkit.org/283950@main">https://commits.webkit.org/283950@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4e77566189a29f888693c8cf63c450e8bf8f76c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67834 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47216 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20473 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71890 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/18975 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69952 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55015 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18782 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54270 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12683 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70901 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43284 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58656 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34737 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39959 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16066 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17333 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61925 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16409 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73587 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11797 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15694 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61719 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11834 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58730 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15079 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9619 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3236 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43023 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/44100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45286 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43838 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -1535,22 +1504,61 @@ ExceptionOr<void> InspectorStyleSheet::setRuleStyleText(const InspectorCSSId& id
+    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(bodyEnd <= styleSheetText.length());
+
+    if (oldRuleText)
+        *oldRuleText = styleSheetText.substring(bodyStart, bodyEnd - bodyStart);
```
should this happen before we call `CSSStyleDeclaration::setCssText`?  i.e. grab the value for both `old*Text` before modifying anything

burg commented on LayoutTests/inspector/css/setStyleText.html:
```diff
@@ -0,0 +1,100 @@
+                await modifyStyleForColor(targetColor);
+            }
+
+            const replacer = (key, value) => {
```
Naming nit: `filteredValueReplacer` ?

burg commented on LayoutTests/inspector/css/setStyleText.html:
```diff
@@ -0,0 +1,100 @@
+            }
+
+            const replacer = (key, value) => {
+                if (key === "ruleId" || key === "styleId" || key === "range" || key === "sourceLine" || key === "sourceURL")
```
```
const filterablePropertyNames = ["ruleId", "styleId", "range", "sourceLine", "sourceURL"];
return (key in filterablePropertyNames) ? "<filtered>" : value;
```

burg commented on LayoutTests/inspector/css/setStyleText.html:
```diff
@@ -0,0 +1,100 @@
+            });
+
+            const getAllAuthorRules = async () => {
+                let { matchedCSSRules: matchedCSSRulesForBody } = await InspectorProtocol.awaitCommand({
```
I would extract this async await into a separate helper function. Then it could be quite straight-line:

```
const getAllAuthorRules = async () => {
return [ getRulesForNode(body), getRulesForNode(element) ]
    .flatMap(...)
    .filter(...)
```

burg commented on LayoutTests/inspector/css/setStyleText.html:
```diff
@@ -0,0 +1,100 @@
+                    .filter((rule) => rule.origin === "author");
+            };
+
+            const modifyStyleForColor = async (targetColor) => {
```
Nit: modifyBackgroundColorStyle

burg commented on LayoutTests/inspector/css/setStyleText.html:
```diff
@@ -0,0 +1,100 @@
+</script>
+</head>
+<body onLoad="runTest()">
+<p>Tests for the  CSS.setStyleText command.</p>
```
Nit: doublespace

burg commented on LayoutTests/inspector/css/setStyleText.html:
```diff
@@ -0,0 +1,100 @@
+};
+</script>
+</head>
+<body onLoad="runTest()">
```
Nit: onload

burg commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -168,12 +168,12 @@ class InspectorCSSAgent::SetStyleTextAction final : public InspectorCSSAgent::St
     ExceptionOr<void> undo() override
     {
-        return m_styleSheet->setRuleStyleText(m_cssId, m_oldText, nullptr, InspectorStyleSheet::IsUndo::Yes);
+        return m_styleSheet->setRuleStyleText(m_cssId, m_oldText, nullptr, &m_oldRuleText, nullptr);
```
This is kind of a mess.

burg commented on Source/WebCore/inspector/InspectorStyleSheet.h:
```diff
@@ -241,7 +239,7 @@ class InspectorStyleSheetForInlineStyle final : public InspectorStyleSheet {
     ExceptionOr<String> text() const final;
     CSSStyleDeclaration* styleForId(const InspectorCSSId& id) const final { ASSERT_UNUSED(id, !id.ordinal()); return &inlineStyle(); }
-    ExceptionOr<void> setRuleStyleText(const InspectorCSSId&, const String& newText, String* oldText, InspectorStyleSheet::IsUndo = InspectorStyleSheet::IsUndo::No) final;
+    ExceptionOr<void> setRuleStyleText(const InspectorCSSId&, const String& newStyleDeclarationText, String* oldStyleDeclarationText, const String*, String*) final;
```
> I'm not sure that these two String* parameters are necessary, or used?

Nevermind

burg commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -1848,17 +1837,17 @@ ExceptionOr<String> InspectorStyleSheetForInlineStyle::text() const
 }
 
-ExceptionOr<void> InspectorStyleSheetForInlineStyle::setRuleStyleText(const InspectorCSSId&, const String& text, String* oldText, IsUndo)
+ExceptionOr<void> InspectorStyleSheetForInlineStyle::setRuleStyleText(const InspectorCSSId&, const String& newStyleDeclarationText, String* oldStyleDeclarationText, const String*, String*)
```
Last two arguments unused, but I think we should still include the names.

burg commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -1537,22 +1506,62 @@ ExceptionOr<void> InspectorStyleSheet::setRuleStyleText(const InspectorCSSId& id
+    return canonicalRuleText;
+}
+
+ExceptionOr<void> InspectorStyleSheet::setRuleStyleText(const InspectorCSSId& id, const String& newStyleDeclarationText, String* oldStyleDeclarationText, const String* newRuleText, String* oldRuleText)
```
Nit: newRuleText should be a const String&
Nit: name the String* arguments to be more obviously out-params. i.e., `returnedOldRuleText`, `out_oldRuleText`, or similar.

the-chenergy replied:
> Nit: newRuleText should be a const String&

~~It could be null for undoing, where we want to avoid computing the canonical rule text. I can add a short comment to signal that.~~

It could be null, for when we want to compute the canonical rule text as the new rule text 😵

burg reviewed (changes requested):
Commit message feedback:
What I wanted to know is why does this change address the problem described. This information is in the radar, so please copy it into the commit message. I think you'll want to describe the overall situation (glitch = bug in the frontend and backend) and that this PR addresses the backend reparsing issue.

Overall this patch looks in good shape, just a few stylistic and convention issues. Please submit another iteration with changes included for EWS testing.

burg reviewed (approved):
Nice work!

the-chenergy committed:
Web Inspector: Glitches when trying to edit a style from a style sheet that has an @import statement
rdar://125185110
https://bugs.webkit.org/show_bug.cgi?id=271403

Reviewed by BJ Burg.

This commit optimizes the handling of CSS.setStyleText by using CSSOM
to update the style declaration text directly, to avoid re-parsing the
entire style sheet.

The observed glitch is a combination of two bugs, both due to having to
re-parse for any style edits:

   1. In the backend, when a style sheet with @import statements is
      re-parsed, it (not just the @import-ed ones) may get removed and
      re-added, causing the CSS agent to mistakenly send
      StyleSheetRemoved and StyleSheetAdded events for that parent
      style sheet. (https://webkit.org/b/279773)

   2. Due to @import-ed style sheets receiving new IDs as a result of
      re-parsing, the frontend thinks any edit to the parent style sheet
      is a "significant change." (Source: https://github.com/WebKit/WebKit/blob/24b5b087d18d4baa487dd73905c46589fa353da3/Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js#L267)
      Handling a significant change triggers a re-render of the entire
      styles panel.

Adjacent to this glitch, there's another notable consequence involving
re-parsing:

   3. Re-parsing an @import-ed style sheet uses its original text from
      its source, meaning editing the parent style sheet will suddenly
      revert all edits made to any @import-ed style sheets.

By avoiding re-parsing for style declaration edits, we fix the immediate
glitch in the simplest fashion and move towards a more efficient
approach to handle editing. This commit is a step in the direction of
eliminating re-parsing for all style editing functions in the inspector:
   1. Editing the style declaration text is addressed by this PR.
   2. Editing the header of a style rule is already done through CSSOM: https://github.com/WebKit/WebKit/blob/6277bccb19de5ebd68797f369f8ff24fc4417f2e/Source/WebCore/inspector/InspectorStyleSheet.cpp#L1116-L1119.
   3. Editing the header of a non-style rule isn't directly supported by
      CSSOM but can be achieved equivalently by removing and
      re-inserting the updated rule using CSSOM.
   4. Adding and 5. removing rules can be optimized to use CSSOM in a
      similar way to this patch.

(Functions 3, 4, and 5 may still malfunction with @import rules after
this patch due to https://webkit.org/b/279773 as they're still done by
re-parsing.)

* Source/WebCore/inspector/InspectorStyleSheet.h:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::computeCanonicalRuleText):
(WebCore::InspectorStyleSheet::setRuleStyleText):
   - Still canonicalize the rule body text to enable showing it in the
     Sources tab, but don't re-parse the whole style sheet just to apply
     the styles. Use CSSOM's support for that.

(WebCore::InspectorStyleSheetForInlineStyle::setRuleStyleText):
   - Adapt to the signature change of setRuleStyleText.

(WebCore::InspectorStyleSheet::styleSheetTextWithChangedStyle): Deleted.
   - Unused.

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
   - Add field to record the old rule body text to enable faithful
     undoing.

* LayoutTests/inspector/css/setStyleText-expected.txt: Added.
* LayoutTests/inspector/css/setStyleText.html: Added.

Canonical link: https://commits.webkit.org/283950@main
